internal/controller/object/object.go

```diff
@@ -483,6 +488,105 @@ func getReferenceInfo(ref v1alpha2.Reference) (string, string, string, string) {
 	return apiVersion, kind, namespace, name
 }
 
+func (c *external) checkDeriveFromObject(observed *unstructured.Unstructured) (ready bool) {
+	conditioned := xpv1.ConditionedStatus{}
+	err := fieldpath.Pave(observed.Object).GetValueInto("status", &conditioned)
+	if err != nil {
+		c.logger.Debug("Got error while getting conditions from observed object, setting it as Unavailable", "error", err, "observed", observed)
+		return
+	}
+	if status := conditioned.GetCondition(xpv1.TypeReady).Status; status != v1.ConditionTrue {
+		c.logger.Debug("Observed object is not ready, setting it as Unavailable", "status", status, "observed", observed)
+		return
+	}
+	ready = true
+	return
+}
+
+func (c *external) checkAllConditions(observed *unstructured.Unstructured) (allTrue bool) {
+	conditioned := xpv1.ConditionedStatus{}
+	err := fieldpath.Pave(observed.Object).GetValueInto("status", &conditioned)
+	if err != nil {
+		c.logger.Debug("Got error while getting conditions from observed object, setting it as Unavailable", "error", err, "observed", observed)
+		return
+	}
+	allTrue = len(conditioned.Conditions) > 0
+	for _, condition := range conditioned.Conditions {
+		if condition.Status != v1.ConditionTrue {
+			allTrue = false
+			break
+		}
+	}
+	return
+}
```

Comment:
nit: we try to avoid naked returns, as they make code more obscure.
```suggestion
func (c *external) checkAllConditions(observed *unstructured.Unstructured) bool {
	conditioned := xpv1.ConditionedStatus{}
	err := fieldpath.Pave(observed.Object).GetValueInto("status", &conditioned)
	if err != nil {
		c.logger.Debug("Got error while getting conditions from observed object, setting it as Unavailable", "error", err, "observed", observed)
		return false
	}
	allTrue = len(conditioned.Conditions) > 0
	for _, condition := range conditioned.Conditions {
		if condition.Status != v1.ConditionTrue {
			return false
		}
	}
	return true
}
```